docs/spec/distributing.rst

```diff
@@ -10,72 +10,428 @@ Stub files
+    _T = TypeVar("_T")
 
-* Just like in `normal Python files <https://docs.python.org/3/reference/import.html#submodules>`_, submodules
-  automatically become exported attributes of their parent module
-  when imported. For example, if the ``spam`` package has the
-  following directory structure::
+    class Foo:
+        def do_things(self: _T) -> _T: ...  # self has type _T
+        @classmethod
+        def create_it(cls: _T) -> _T: ...  # cls has type _T
 
-      spam/
-          __init__.pyi
-          ham.pyi
+Using a function or method body other than the ellipsis literal is currently
+unspecified. Stub authors may experiment with other bodies, but it is up to
+individual type checkers how to interpret them::
 
-  where ``__init__.pyi`` contains a line such as ``from . import ham``
-  or ``from .ham import Ham``, then ``ham`` is an exported attribute
-  of ``spam``.
+    def foo(): ...  # compatible
+    def bar(): pass  # behavior undefined
 
-* Stub files may be incomplete. To make type checkers aware of this, the file
-  can contain the following code::
+All variants of overloaded functions and methods must have an ``@overload``
+decorator::
 
-    def __getattr__(name) -> Any: ...
+    @overload
+    def foo(x: str) -> str: ...
+    @overload
+    def foo(x: float) -> int: ...
 
-  Any identifier not defined in the stub is therefore assumed to be of type
-  ``Any``.
+The following (which would be used in the implementation) is wrong in stubs::
+
+    @overload
+    def foo(x: str) -> str: ...
+    @overload
+    def foo(x: float) -> int: ...
+    def foo(x: str | float) -> Any: ...
+
+Aliases and NewType
+"""""""""""""""""""
+
+Type checkers should accept module-level type aliases, optionally using
+``TypeAlias`` (:pep:`613`), e.g.::
+
+  _IntList = list[int]
+  _StrList: TypeAlias = list[str]
+
+Type checkers should also accept regular module-level or class-level aliases,
+e.g.::
+
+  def a() -> None: ...
+  b = a
+
+  class C:
+      def f(self) -> int: ...
+      g = f
+
+A type alias may contain type variables. As per :pep:`484`,
+all type variables must be substituted when the alias is used::
+
+  _K = TypeVar("_K")
+  _V = TypeVar("_V")
+  _MyMap: TypeAlias = dict[str, dict[_K, _V]]
+
+  # either concrete types or other type variables can be substituted
+  def f(x: _MyMap[str, _V]) -> _V: ...
+  # explicitly substitute in Any rather than using a bare alias
+  def g(x: _MyMap[Any, Any]) -> Any: ...
+
+Otherwise, type variables in aliases follow the same rules as type variables in
+generic class definitions.
+
+``typing.NewType`` is also supported in stubs.
+
+Decorators
+""""""""""
+
+Type checkers are expected to understand the effects of all decorators defined
+in the ``typing`` module, plus these additional ones:
+
+ * ``classmethod``
+ * ``staticmethod``
+ * ``property`` (including ``.setter``)
+ * ``abc.abstractmethod``
+ * ``dataclasses.dataclass``
+ * functions decorated with ``@typing.dataclass_transform``
```

Comment:
Plus classes decorated with `@typing.dataclass_transform`.

Reply:
I thought that classes decorated with dataclass_transform were inherited from, not used as decorators?

Reply:
That's correct, but the `dataclass_transform` is used then as a decorator on the metaclass or base class. This class decorator must be supported by type checkers. The current wording above says "functions decorated with `@typing.dataclass_transform`".

Reply:
That's covered by "all decorators defined in the `typing` module" (in the paragraph above this list).

Reply:
I also had this question while reading this section. Probably

```
* functions decorated with ``@typing.dataclass_transform``
``` 

can be deleted, since 

```
Type checkers are expected to understand the effects of all decorators defined
in the ``typing`` module, plus these additional ones:
```

covers it.

Reply:
I think the reasoning behind the inclusion of this bullet, with its current wording, is that a function decorated with `typing.dataclass_transform` is itself a decorator (which is not itself defined in the `typing` module), whose function must be understood by type checkers.

A class decorated by `typing.dataclass_transform` is not then itself a decorator, so this case is fully covered by "all decorators defined in the `typing` module."